pkg/roachprod/failureinjection/failures/disk_stall.go

```diff
@@ -0,0 +1,474 @@
+
+package failures
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/roachprod"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/logger"
+	"github.com/cockroachdb/errors"
+)
+
+var cockroachIOController = filepath.Join("/sys/fs/cgroup/system.slice", install.VirtualClusterLabel(install.SystemInterfaceName, 0)+".service", "io.max")
+
+type CGroupDiskStaller struct {
+	GenericFailure
+}
+
+func MakeCgroupDiskStaller(clusterName string, l *logger.Logger, secure bool) (FailureMode, error) {
+	c, err := roachprod.GetClusterFromCache(l, clusterName, install.SecureOption(secure))
+	if err != nil {
+		return nil, err
+	}
+	genericFailure := GenericFailure{c: c, runTitle: CgroupsDiskStallName}
+	return &CGroupDiskStaller{GenericFailure: genericFailure}, nil
+}
+
+const CgroupsDiskStallName = "cgroup-disk-stall"
+
+func registerCgroupDiskStall(r *FailureRegistry) {
+	r.add(CgroupsDiskStallName, DiskStallArgs{}, MakeCgroupDiskStaller)
+}
+
+func (s *CGroupDiskStaller) Description() string {
+	return CgroupsDiskStallName
+}
+
+type DiskStallArgs struct {
+	StallLogs   bool
+	StallReads  bool
+	StallWrites bool
+	// If true, allow the failure mode to restart nodes as needed. E.g. dmsetup requires
+	// the cockroach process to not be running to properly setup. If RestartNodes is true,
+	// then the failure mode will restart the cluster for the user.
+	RestartNodes bool
+	// Throughput is the bytes per second to throttle reads/writes to. If unset, will
+	// stall reads/write completely. Supported only for cgroup disk staller, dmsetup
+	// only supports fully stalling reads/writes.
+	Throughput int
+	Nodes      install.Nodes
+}
+
+func (s *CGroupDiskStaller) Setup(ctx context.Context, l *logger.Logger, args FailureArgs) error {
+	diskStallArgs := args.(DiskStallArgs)
+
+	if diskStallArgs.StallLogs {
+		if err := s.Run(ctx, l, s.c.Nodes, "mkdir -p {store-dir}/logs"); err != nil {
+			return err
+		}
+		if err := s.Run(ctx, l, s.c.Nodes, "rm -f logs && ln -s {store-dir}/logs logs || true"); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+func (s *CGroupDiskStaller) Cleanup(ctx context.Context, l *logger.Logger, args FailureArgs) error {
+	stallType := []bandwidthType{readBandwidth, writeBandwidth}
+	nodes := args.(DiskStallArgs).Nodes
+
+	// Setting cgroup limits is idempotent so attempt to unlimit reads/writes in case
+	// something went wrong in Recover.
+	err := s.setThroughput(ctx, l, stallType, throughput{limited: false}, nodes, cockroachIOController)
+	if err != nil {
+		l.PrintfCtx(ctx, "error unstalling the disk; stumbling on: %v", err)
+	}
+	return nil
+}
+
+func getStallTypes(diskStallArgs DiskStallArgs) ([]bandwidthType, error) {
+	var stallTypes []bandwidthType
+	if diskStallArgs.StallWrites {
+		stallTypes = []bandwidthType{writeBandwidth}
+	}
+	if diskStallArgs.StallReads {
```

Comment:
Should we further validate if `StallLogs` is set _without_ `StallWrites`? In that case, we don't expect the process to panic.